synapse_user_restrictions/config.py

```diff
@@ -0,0 +1,127 @@
+import abc
+import enum
+import re
+from enum import Enum
+from typing import Any, List, Pattern, Set, Type
+
+import attr
+import cattr
+from attr import Attribute
+from cattr import Converter
+
+
+def make_cattr_converter() -> Converter:
+    """
+    Creates a cattr `Converter` instance, with custom registered hooks for
+    regular expressions.
+    """
+    converter = cattr.Converter()
+
+    def regex_structure_hook(value: Any, _kind: Type[Pattern[str]]) -> Pattern[str]:
+        if not isinstance(value, str):
+            raise ValueError("Regex patterns should be specified as strings.")
+
+        try:
+            return re.compile(value)
+        except re.error as e:
+            raise ValueError(f"Invalid regex '{value}': {e.msg}")
+
+    converter.register_structure_hook(Pattern[str], regex_structure_hook)
+
+    return converter
+
+
+def check_all_permissions_understood(
+    _instance: Any, _attribute: "Attribute[Set[str]]", permissions: Set[str]
+) -> None:
+    """
+    Checks that all the permissions contained in the list of permissions are
+    ones that we understand and recognise.
+    """
+    for permission in permissions:
+        if permission not in ALL_UNDERSTOOD_PERMISSIONS:
+            nice_list_of_understood_permissions = ", ".join(
+                sorted(ALL_UNDERSTOOD_PERMISSIONS)
+            )
+            raise ValueError(
+                f"{permission!r} is not a permission recognised "
+                f"by the User Restrictions module; "
+                f"try one of: {nice_list_of_understood_permissions}"
+            )
+
+
+class RuleResult(Enum):
+    NoDecision = enum.auto()
+    Allow = enum.auto()
+    Deny = enum.auto()
+
+
+class Rule(abc.ABC):
```

Comment:
This seems unnecessary since we only have one type of rule? I'd remove it until we need it.

Reply:
Yeah OK, it just seemed like good practice to define the boundary between what a specific rule is and what a generic rule is, but can cut it out